Reduce use of memcpy() / memset() in WTF/

cdumez opened the pull request:
#### fecc646ca0f6dc15715760d970795164f6af9106
<pre>
Reduce use of memcpy() / memset() in WTF/
<a href="https://bugs.webkit.org/show_bug.cgi?id=285004">https://bugs.webkit.org/show_bug.cgi?id=285004</a>

Reviewed by Geoffrey Garen and Darin Adler.

* Source/WTF/wtf/Assertions.cpp:
* Source/WTF/wtf/BitVector.cpp:
(WTF::BitVector::setSlow):
(WTF::BitVector::clearAll):
(WTF::BitVector::resizeOutOfLine):
* Source/WTF/wtf/BitVector.h:
* Source/WTF/wtf/Brigand.h:
(brigand::real_::operator value_type const):
* Source/WTF/wtf/ConcurrentBuffer.h:
* Source/WTF/wtf/CurrentTime.cpp:
(WTF::lowResUTCTime):
* Source/WTF/wtf/FastBitVector.cpp:
(WTF::FastBitVectorWordOwner::setEqualsSlow):
(WTF::FastBitVectorWordOwner::resizeSlow):
* Source/WTF/wtf/FastBitVector.h:
(WTF::FastBitVectorWordOwner::operator=):
(WTF::FastBitVectorWordOwner::setAll):
(WTF::FastBitVectorWordOwner::clearAll):
(WTF::FastBitVectorWordOwner::set):
(WTF::FastBitVectorWordOwner::wordsSpan):
(WTF::FastBitVectorWordOwner::wordsSpan const):
* Source/WTF/wtf/Packed.h:
(WTF::PackedAlignedPtr::get const):
(WTF::PackedAlignedPtr::set):
* Source/WTF/wtf/SmallSet.h:
* Source/WTF/wtf/StdLibExtras.h:
(WTF::memcpySpan):
* Source/WTF/wtf/UUID.h:
(WTF::UUID::UUID):
* Source/WTF/wtf/UnalignedAccess.h:
(WTF::unalignedLoad):
(WTF::unalignedStore):
* Source/WTF/wtf/cocoa/FileSystemCocoa.mm:
(WTF::FileSystemImpl::createTemporaryDirectory):
* Source/WTF/wtf/text/StringImpl.h:
(WTF::StringImpl::tryGetUTF8ForCharacters):
* Source/WTF/wtf/text/cf/StringConcatenateCF.h:
(WTF::StringTypeAdapter&lt;CFStringRef&gt;::writeTo&lt;LChar&gt; const):
* Source/WTF/wtf/text/icu/TextBreakIteratorICU.h:
(WTF::TextBreakIteratorICU::makeLocaleWithBreakKeyword):

Canonical link: <a href="https://commits.webkit.org/288265@main">https://commits.webkit.org/288265@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/1eebbd94115751e0090424509a0c1ed246052441

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/82286 "4 style errors") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/1950 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/36354 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/87418 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/33347 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/84391 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/2021 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/9832 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/64130 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/21881 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/85355 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/1409 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/74873 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/44408 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/1311 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/29054 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/32388 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/75272 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/72605 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/29678 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/88774 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/81340 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/9592 "Built successfully") | [❌ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/6850 "Found 3 new test failures: http/tests/download/anchor-download-redirect-cross-origin-top-level.html http/tests/webgpu/webgpu/api/validation/queue/destroyed/query_set.html media/video-replaces-poster.html (failure)") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/72528 "Passed tests") | 
| [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🧪 services](https://ews-build.webkit.org/#/builders/28/builds/81923 "Build is in progress. Recent messages:") | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/9818 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/70687 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/71746 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/15878 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/14885 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/947 "Built successfully") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/12774 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/9545 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/15066 "Built successfully") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/103753 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/9419 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/25170 "Passed tests") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/12885 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/11189 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

darinadler commented on Source/WTF/wtf/BitVector.cpp:
```diff
@@ -232,32 +232,27 @@ bool BitVector::equalsSlowCaseFast(const BitVector& other) const
+    for (size_t i = 0; i < commonWordCount; ++i) {
+        if (myWords[i] != otherWords[i])
             return false;
     }
```
Use `equalSpan` and `first` instead?

darinadler commented on Source/WTF/wtf/BitVector.h:
```diff
@@ -469,7 +469,12 @@ class BitVector final {
         size_t numWords() const { return (m_numBits + bitsInPointer() - 1) / bitsInPointer(); }
         uintptr_t* bits() { return std::bit_cast<uintptr_t*>(this + 1); }
         const uintptr_t* bits() const { return std::bit_cast<const uintptr_t*>(this + 1); }
```
Do these still need to exist? Do they need to be public?

darinadler commented on Source/WTF/wtf/text/cf/StringConcatenateCF.h:
```diff
@@ -50,7 +51,7 @@ inline StringTypeAdapter<CFStringRef>::StringTypeAdapter(CFStringRef string)
 {
     if (m_string)
-        std::memcpy(destination.data(), CFStringGetCStringPtr(m_string, kCFStringEncodingISOLatin1), CFStringGetLength(m_string));
+        memcpySpan(destination, unsafeMakeSpan(CFStringGetCStringPtr(m_string, kCFStringEncodingISOLatin1), CFStringGetLength(m_string)));
```
Should the `CFStringGetCStringPtr` wrapped to return a span be a separate function rather than a subexpression? Maybe it can eventually be shared? I guess it's inconvenient in a header like this.

darinadler commented on Source/WTF/wtf/cf/VectorCF.h:
```diff
@@ -160,6 +160,14 @@ template<typename MapLambdaType> Vector<typename LambdaTypeTraits<MapLambdaType>
+    auto* characters = CFStringGetCStringPtr(string, encoding);
+    if (!characters)
+        return { };
+    return unsafeMakeSpan(characters, CFStringGetLength(string));
```
I don't think this length is right. The C string length is based on the encoding passed and won't necessarily match this length, which is the number of UTF-16 code units.

geoffreygaren replied:
One way to make this correct would be to name it CFStringGetASCIICStringSpan and force kCFStringEncodingASCII, like our caller used to do. (In ASCII, the number of code units is equal to the number characters.)

cdumez replied:
At least one of the call sites was using `kCFStringEncodingISOLatin1`

cdumez replied:
I think I can just call `strlen()`? Or even `span(const char*)` (which ends up calling strlen())

geoffreygaren replied:
Latin1 is still ok (1 byte per character). So you could offer an enum to pick ASCII or Latin1, or two separate functions. 

I don’t think strlen is ok because:
* if a future caller passes a multi-byte encoding, you might find a null byte before end of string
* kind of a bummer to pay the cost of enumeration when you don’t need to

cdumez replied:
The doc for CFStringGetCStringPtr clearly states:
> The encoding must specify an 8-bit encoding.

c.f. https://developer.apple.com/documentation/corefoundation/1542133-cfstringgetcstringptr?language=objc

It returns a C String, which is a null terminated string.

cdumez replied:
I agree though that `strlen()` adds a cost. I just don't think we need to worry about multi-byte encoding since they are not supported by this API (I can add a comment to this effect for clarity).

cdumez replied:
Anyway, to avoid the cost of strlen(), I guess your proposal is still best: Add a enum parameter to allow requesting either ASCII or Latin1 (only).

cdumez replied:
I went with one of Geoff's proposal, separate CFStringGetASCIICStringSpan / CFStringGetLatin1CStringSpan functions and using `CFStringGetLength()`

darinadler reviewed (approved):
I think it's fine to either use the strlen version or the "only Latin-1 or ASCII" version.

cdumez committed:
Reduce use of memcpy() / memset() in WTF/
https://bugs.webkit.org/show_bug.cgi?id=285004

Reviewed by Geoffrey Garen and Darin Adler.

* Source/WTF/wtf/Assertions.cpp:
* Source/WTF/wtf/BitVector.cpp:
(WTF::BitVector::setSlow):
(WTF::BitVector::clearAll):
(WTF::BitVector::resizeOutOfLine):
* Source/WTF/wtf/BitVector.h:
* Source/WTF/wtf/Brigand.h:
(brigand::real_::operator value_type const):
* Source/WTF/wtf/ConcurrentBuffer.h:
* Source/WTF/wtf/CurrentTime.cpp:
(WTF::lowResUTCTime):
* Source/WTF/wtf/FastBitVector.cpp:
(WTF::FastBitVectorWordOwner::setEqualsSlow):
(WTF::FastBitVectorWordOwner::resizeSlow):
* Source/WTF/wtf/FastBitVector.h:
(WTF::FastBitVectorWordOwner::operator=):
(WTF::FastBitVectorWordOwner::setAll):
(WTF::FastBitVectorWordOwner::clearAll):
(WTF::FastBitVectorWordOwner::set):
(WTF::FastBitVectorWordOwner::wordsSpan):
(WTF::FastBitVectorWordOwner::wordsSpan const):
* Source/WTF/wtf/Packed.h:
(WTF::PackedAlignedPtr::get const):
(WTF::PackedAlignedPtr::set):
* Source/WTF/wtf/SmallSet.h:
* Source/WTF/wtf/StdLibExtras.h:
(WTF::memcpySpan):
* Source/WTF/wtf/UUID.h:
(WTF::UUID::UUID):
* Source/WTF/wtf/UnalignedAccess.h:
(WTF::unalignedLoad):
(WTF::unalignedStore):
* Source/WTF/wtf/cocoa/FileSystemCocoa.mm:
(WTF::FileSystemImpl::createTemporaryDirectory):
* Source/WTF/wtf/text/StringImpl.h:
(WTF::StringImpl::tryGetUTF8ForCharacters):
* Source/WTF/wtf/text/cf/StringConcatenateCF.h:
(WTF::StringTypeAdapter<CFStringRef>::writeTo<LChar> const):
* Source/WTF/wtf/text/icu/TextBreakIteratorICU.h:
(WTF::TextBreakIteratorICU::makeLocaleWithBreakKeyword):

Canonical link: https://commits.webkit.org/288265@main
